tools/rtpsrelay/GuidAddrSet.cpp

```diff
@@ -7,6 +7,103 @@
 
 namespace RtpsRelay {
 
+bool AddrSetStats::upsert_address(const AddrPort& remote_address,
+                      const OpenDDS::DCPS::MonotonicTimePoint& now,
+                      const OpenDDS::DCPS::MonotonicTimePoint& expiration,
+                      size_t max_ip_addresses)
+{
+  ACE_INET_Addr addr_only(remote_address.addr);
+  addr_only.set_port_number(0);
+  auto iter = ip_to_ports.find(addr_only);
+  if (iter == ip_to_ports.end()) {
+    if (max_ip_addresses > 0 && ip_to_ports.size() == max_ip_addresses) {
+      return false;
+    }
+    iter = ip_to_ports.insert(std::make_pair(addr_only, PortSet())).first;
+  }
+
+  relay_stats_reporter_.max_ips_per_client(static_cast<uint32_t>(ip_to_ports.size()), now);
+
+  std::map<u_short, OpenDDS::DCPS::MonotonicTimePoint>* port_map = nullptr;
+  switch (remote_address.port) {
+  case SPDP:
+    port_map = &iter->second.spdp_ports;
+    break;
+  case SEDP:
+    port_map = &iter->second.sedp_ports;
+    break;
+  case DATA:
+    port_map = &iter->second.data_ports;
+    break;
+  }
+  if (!port_map) {
+    return false;
+  }
+  const auto pair = port_map->insert(std::make_pair(remote_address.addr.get_port_number(), expiration));
+  if (pair.second) {
+    return true;
+  }
+  pair.first->second = expiration;
+  return false;
+}
+
+bool AddrSetStats::remove_if_expired(const AddrPort& remote_address, const OpenDDS::DCPS::MonotonicTimePoint& now,
+                                     bool& ip_now_unused, OpenDDS::DCPS::MonotonicTimePoint& updated_expiration)
+{
+  ACE_INET_Addr addr_only(remote_address.addr);
+  addr_only.set_port_number(0);
+  const auto iter = ip_to_ports.find(addr_only);
+  if (iter == ip_to_ports.end()) {
+    return false;
+  }
+
+  std::map<u_short, OpenDDS::DCPS::MonotonicTimePoint>* port_map = nullptr;
```

Comment:
Maybe make a `typedef`/`using` of this type to share with `PortSet`? Could also make the switch statement a method in `PortSet`.